lib/ansible/modules/blockinfile.py

```diff
@@ -336,7 +357,26 @@ def main():
         if not lines[n0 - 1].endswith(b(os.linesep)):
             lines[n0 - 1] += b(os.linesep)
 
+    # Before the block: check if we need to prepend a blank line
+    # If yes, we need to add the blank line if we are at the beginning of the file
+    # or if the previous line is not a blank line
+    # In both cases, we need to shift by one on the right the inserting position of the block
+    if params['prepend_newline'] and present:
+        if n0 == 0 or lines[n0 - 1] != b(os.linesep):
```

Comment:
Looks good to me, though I wondered about one thing - https://github.com/ansible/ansible/issues/80835 suggests excluding inserting newlines at the start or end of the file. That makes sense to me and manually adding newlines at the beginning/end of the file would not difficult. What about
```python
        if n0 != 0 and lines[n0 - 1] != b(os.linesep):
```